enhancements/operator-sdk/downstream-operator-sdk.md

```diff
@@ -0,0 +1,385 @@
+
+### Risks and Mitigation
+
+Moving forward with the automated syncing is a little high risk in the sense
+that there are some unknown issues we may encounter. We will mitigate this risk
+by keeping the existing `openshift/ocp-release-operator-sdk` repo until the
+automated `openshift/operator-sdk` repo is working. Once the cut over is made we
+can remove the `openshift/ocp-release-operator-sdk` repo.
+
+In the event that the automated repo is not ready in time for 4.7, the backup
+plan is to do a 4.7 release of OperatorSDK from the existing
+`openshift/ocp-release-operator-sdk` repo.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+The existing set of Operator SDK tests will be run downstream. Today we run
+tests for the Ansible and Helm operators.
+
+### Graduation Criteria
+
+Operator SDK will go out once the downstream builds work. The Operator SDK will
+be downloadable from the [Red Hat Developer page][rh-dev-page].
+
+### Upgrade / Downgrade Strategy
+
+All Operator SDK releases have a migration guide on what users have to do to
+upgrade their operators. These guides will likely be made available in the
+official OpenShift documentation.
+
+### Version Skew Strategy
+
+* The operator-sdk repo will vendor its dependencies to keep consistent builds
+* The ansible and helm repos will also vendor any Go dependencies
+* The operator-lib repo will just be a mirror of the upstream
+
+#### Branching
+
+As described in the [Repos](#repos) and [Syncing](#syncing) sections, the
+downstream repos will use a different branching strategy.
+
+The downstream repos will contain mirrors of the upstream repos, an overlay
+branch, and release branches that correspond to OpenShift releases.
+
+* master - mirrors upstream master
+* v1.1.x - mirrors upstream v1.1.x
+* v1.0.x - mirrors upstream v1.0.x
+* v0.19.x - mirrors upstream v0.19.x
+* overlay - contains the downstream [overlay files][overlay-files]
+* release-4.8 - SDK release + overlay for OpenShift 4.8
+* release-4.7 - SDK release + overlay for OpenShift 4.7
+
+The release branches will have specific tagged versions of the SDK only. We will
+not pull in all commits from an upstream release branch unless they have been
+tagged and released upstream.
+
+The advantage of managing our own branches is it gives us flexibility in
+keeping master in sync with upstream master so we always have the latest code
+available downstream.
+
+The disadvantage to this is that each release branch is not managed by ART.
+
+## Implementation History
+
+2020-10-13 - Rewrite proposal using the automation option.
+2020-10-07 - Propose to automate the syncing of upstream to downstream.
+2020-09-10 - Propose using the UPSTREAM-MERGE.sh script similar to Service
+             Catalog releases.
+
+## Drawbacks
+
+* We would have to manage our own downstream release branches
+* We would have to handle the automation of updates ourselves
+* The special downstream branch is kind of janky
```

Comment:
If we put the downstream-only files in a separate directory that will never exist in the upstream repo, doesn't that mean we'll never have merge conflicts when we sync upstream to downstream?

I know @fabianvf has probably thought through all of this in more excruciating detail than I have, but refresh my memory on why we can't just commit the downstream-only files in downstream master?

Reply:
The downstream-only files weren't usually the source of major conflicts. From what I've seen it's been the slow sync rate we had so we would go from 0.15 to 0.17 to 0.19 into the master branch. I tried bringing in v1.0.0 to current ocp-release-operator-sdk there are a ton of conflicts with the main files. Mostly because things moved into the internal directory.

So the idea with the automated version is the branches (including master) are direct mirrors of upstream. The release branches will be a tagged version + overlay.

Reply:
> I tried bringing in v1.0.0 to current ocp-release-operator-sdk there are a ton of conflicts with the main files. Mostly because things moved into the internal directory.

I _think_ (could be wrong) that this is because of the history in how ocp-release-operator-sdk was created and originally setup. Things _should_ merge cleanly if upstream and downstream start from the same base, even if extra downstream-only files exist downstream.

Conceptually, what I'm wondering is:

If we never touch any of the files from upstream in our downstream branch, upstream will always merge cleanly into downstream (right?). Then, instead of maintaining a separate overlay branch and dealing with merging that into new downstream branches, could we just put the overlay branch contents directly on `master` in a separate downstream-only directory containing our downstream-only Dockerfiles, Makefiles, patches, etc?

Whenever we branch from master downstream, we would get the downstream-only directory contents for free because they would already be on `master` rather than in some other branch that we have to merge in.